### PR TITLE
feat: RND-119: Add token usage in inference output

### PR DESCRIPTION
**Pull request by niklub**
Adala table now contains the following system fields:
- `_prompt_tokens`
- `_completion_tokens`
- `_total_cost_usd`

and they're piped to the LSE resulthandler.

TODO need to update all tests again or add some kind of flag which only adds these fields when requested by server

### CODE REVIEW

**Comment by robot-ci-heartex** [comment]
Commit https://github.com/HumanSignal/infra/commit/d5ee8986f9b410ce707324493bb7be52b2eca902 is created.

**Comment by robot-ci-heartex** [comment]
Commit https://github.com/HumanSignal/infra/commit/b0aa698fdcb3db0e5d8a076a3bf1e1dc4c37fa8a is created.

File: adala/runtimes/_litellm.py [hunk 1]
```diff
@@ -191,27 +213,43 @@ def record_to_record(
...
                 **self.model_extra,
             )
+            usage = completion.usage
+            dct = response.dict()
+        except IncompleteOutputException as e:
+            usage = e.total_usage
+            dct = _format_error_dict(e)
+            print_error(f"Inference error {dct['_adala_message']}: Incomplete output. Increase max_tokens.")
+            tb = traceback.format_exc()
+            logger.error(tb)
```
Comment (niklub) [inline adala/runtimes/_litellm.py:223]: to improve code redability, let's move these blocks in a separate function, and probably we don't need double logging:
```
def _log_llm_exception(self, e):
     dct = _format_error_dict(e)
     base_error = f"Inference error {dct['_adala_message']}: Incomplete output. Increase max_tokens."
     tb = traceback.format_exc()
     logger.error(f'{base_error}\nTraceback:\n{tb}')
```

File: server/handlers/result_handlers.py [hunk 2]
```diff
@@ -80,6 +80,11 @@ class LSEBatchItem(BaseModel):
     message: Optional[str] = Field(None, alias="_adala_message")
     details: Optional[str] = Field(None, alias="_adala_details")
 
+    prompt_tokens : int = Field(alias="_prompt_tokens")
+    completion_tokens : int = Field(alias="_completion_tokens")
+    # this can fail to calculate
+    total_cost_usd : Optional[float] = Field(alias="_total_cost_usd")
```
Comment (niklub) [inline server/handlers/result_handlers.py:86]: let's keep prompt cost and completion cost as well

File: adala/runtimes/_litellm.py [hunk 3]
```diff
@@ -58,6 +70,19 @@ def _format_error_dict(e: Exception) -> dict:
...
+def _update_with_usage(data: Dict, usage: Usage, model: str) -> None:
+    data["_prompt_tokens"] = usage.prompt_tokens
+    # will not exist if there is no completion
+    data["_completion_tokens"] = usage.get('completion_tokens', 0)
+    # can't use litellm.completion_cost bc it only takes the most recent completion, and .usage is summed over retries
+    # TODO make sure this is calculated correctly after we turn on caching
+    # litellm will register the cost of an azure model on first successful completion. If there hasn't been a successful completion, the model will not be registered
+    try:
+        prompt_cost, completion_cost = litellm.cost_per_token(model, usage.prompt_tokens, usage.get('completion_tokens', 0))
+        data["_total_cost_usd"] = prompt_cost + completion_cost
```
[the remark this answers is not in the document]
Comment (matt-bernstein) [reply adala/runtimes/_litellm.py:82]: can't do any nesting bc we're still going to/from a dataframe row. that really needs to be fixed at some point but don't want to block this pr

**Review by hakan458 — APPROVED** [review]

**Comment by robot-ci-heartex** [comment]
Commit https://github.com/HumanSignal/infra/commit/61e89c98abfb515fdd87e17e8e921b48cc193cf0 is created.

**Comment by robot-ci-heartex** [comment]
Commit https://github.com/HumanSignal/infra/commit/4fe5d1f6a1819db91e4c38718876a71f4ad893eb is created.

**Comment by robot-ci-heartex** [comment]
Commit https://github.com/HumanSignal/infra/commit/eca9fec610779a55a044929e0b3bfe6f2550a7ba is created.

**Comment by robot-ci-heartex** [comment]
Commit https://github.com/HumanSignal/infra/commit/d4bdadffdda7d05911b0b1e722ebe66029891443 is created.

**Pull request merged**